<a href="https://colab.research.google.com/github/venezianof/booksum/blob/main/openmed.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [8]:
def search_cardiac_tumors(example):
    for msg in example['messages']:
        text = msg['content'].lower()
        if 'angiosarcoma' in text or 'tumore cardiaco' in text:
            return True
    return False

# Search in the broad dataset
tumor_examples = dataset['train'].filter(search_cardiac_tumors)

print(f'Trovati {len(tumor_examples)} esempi relativi ai tumori cardiaci o angiosarcoma.')

if len(tumor_examples) > 0:
    display(tumor_examples[0])

Filter:   0%|          | 0/810284 [00:00<?, ? examples/s]

Trovati 176 esempi relativi ai tumori cardiaci o angiosarcoma.


{'messages': [{'content': 'Does canine hemangiosarcoma originate from hematopoietic precursors with potential for endothelial differentiation?\n\nA.: yes\nB.: no\nC.: maybe\n\nPlease reason step by step, and put your final answer within \\boxed{}.',
   'reasoning_content': None,
   'role': 'user'},
  {'content': 'Canine hemangiosarcoma is a malignant tumor that primarily arises from endothelial cells, which line the blood vessels. Endothelial cells are derived from mesodermal precursors, specifically angioblasts, and are distinct from hematopoietic precursors. Hematopoietic precursors, such as hematopoietic stem cells, give rise to blood cells (e.g., red blood cells, white blood cells, platelets) and are associated with hematopoietic malignancies like leukemia or lymphoma. While both endothelial and hematopoietic cells originate from mesoderm, they represent separate lineages with different differentiation potentials. Hemangiosarcoma is characterized by its vascular origin and is not c

In [7]:
def specifically_popeye(example):
    for msg in example['messages']:
        if 'popeye' in msg['content'].lower():
            return True
    return False

# Filter the previous biceps_examples for 'popeye'
popeye_subset = biceps_examples.filter(specifically_popeye)

print(f'Trovati {len(popeye_subset)} esempi specifici per il segno di Popeye.')

# Analyze the first 3 examples to identify common clinical descriptions
for i in range(min(3, len(popeye_subset))):
    print(f'\n--- Analisi Esempio {i+1} ---')
    print(f"Q: {popeye_subset[i]['messages'][0]['content'][:200]}...")
    # Display the reasoning if available
    reasoning = popeye_subset[i]['messages'][1].get('reasoning_content')
    if reasoning:
        print(f"Ragionamento del Modello: {reasoning[:500]}...")
    else:
        print(f"Risposta: {popeye_subset[i]['messages'][1]['content'][:500]}...")

Filter:   0%|          | 0/858 [00:00<?, ? examples/s]

Trovati 3 esempi specifici per il segno di Popeye.

--- Analisi Esempio 1 ---
Q: Is spinach exceptionally high in iron compared to other vegetables?...
Ragionamento del Modello: Okay, the user is asking if spinach is exceptionally high in iron compared to other vegetables. Let me start by recalling what I know about spinach's iron content. I remember that spinach is often touted as a great source of iron, especially in popular culture like Popeye. But I need to check the actual nutritional data.

First, I should compare spinach's iron content per 100 grams to other vegetables. Let me think about legumes like lentils or beans. They usually have higher iron. Also, leafy g...

--- Analisi Esempio 2 ---
Q: Do patients Have Strong Preferences and Perceptions for Biceps Tenotomy Versus Tenodesis?

A.: yes
B.: no
C.: maybe

Please reason step by step, and put your final answer within \boxed{}....
Ragionamento del Modello: First, the question is: "Do patients have strong preferences and percep

In [6]:
def search_biceps_sign(example):
    for msg in example['messages']:
        text = msg['content'].lower()
        if 'bicipite' in text or 'biceps' in text or 'popeye' in text:
            return True
    return False

# Search in the broad dataset for clinical accuracy
biceps_examples = dataset['train'].filter(search_biceps_sign)

print(f'Trovati {len(biceps_examples)} esempi relativi al bicipite o al segno cercato.')

if len(biceps_examples) > 0:
    display(biceps_examples[0])

Filter:   0%|          | 0/810284 [00:00<?, ? examples/s]

Trovati 858 esempi relativi al bicipite o al segno cercato.


{'messages': [{'content': 'Which is the nerve root of Biceps jerk?\na. C5-C6\nb. C6-C8\nc. C7-C8\nd. C8-T1\n\nPlease reason step by step, and put your final answer within \\boxed{}.',
   'reasoning_content': None,
   'role': 'user'},
  {'content': 'The Biceps jerk reflex is a deep tendon reflex test that assesses the integrity of the C5-C6 nerve roots. When the biceps tendon is tapped, it stretches the biceps muscle, activating muscle spindles. Sensory signals travel via the musculocutaneous nerve (derived from C5 and C6 nerve roots) to the spinal cord, and motor signals return via the same nerve, causing contraction. \n\n- Option a (C5-C6) is correct, as the Biceps reflex is specifically associated with these nerve roots.\n- Option b (C6-C8) corresponds to the Triceps reflex, not the Biceps reflex.\n- Option c (C7-C8) is also associated with the Triceps reflex.\n- Option d (C8-T1) relates to the finger jerk reflex, involving the ulnar nerve.\n\nThus, the nerve root for the Biceps jerk

In [3]:
def contains_diabetes(example):
    # Check for 'diabetes' or 'diabete' (Italian) in the conversation history
    for msg in example['messages']:
        text = msg['content'].lower()
        if 'diabetes' in text or 'diabete' in text:
            return True
    return False

# Filter the dataset
diabetes_subset = dataset['train'].filter(contains_diabetes)

print(f'Trovati {len(diabetes_subset)} esempi relativi al diabete.')

# Mostriamo il primo esempio trovato
if len(diabetes_subset) > 0:
    display(diabetes_subset[0])

Filter:   0%|          | 0/810284 [00:00<?, ? examples/s]

Trovati 36042 esempi relativi al diabete.


{'messages': [{'content': 'Do advanced glycation end products inhibit the expression of collagens type I and III by human gingival fibroblasts?\n\nA.: yes\nB.: no\nC.: maybe\n\nPlease reason step by step, and put your final answer within \\boxed{}.',
   'reasoning_content': None,
   'role': 'user'},
  {'content': 'The question asks whether advanced glycation end products (AGEs) inhibit the expression of collagen type I and III by human gingival fibroblasts. \n\nAGEs are known to accumulate in tissues under conditions such as diabetes and aging, and they can bind to receptors like RAGE (Receptor for AGEs) on cells, including fibroblasts. This binding activates signaling pathways that often lead to reduced collagen synthesis. Specifically, studies on human gingival fibroblasts have demonstrated that AGEs downregulate the expression of collagen genes (e.g., COL1A1 and COL3A1), resulting in decreased production of collagen type I and III. For instance, research shows that AGEs inhibit coll

### Come utilizzare queste informazioni

Se sospetti di avere sintomi legati al diabete, ecco alcuni passi consigliati:

1.  **Monitora i tuoi sintomi:** Prendi nota della frequenza e dell'intensità di ciò che provi (ad esempio, quante volte ti svegli di notte per urinare).
2.  **Prepara una lista di domande:** Prima di andare dal medico, scrivi i tuoi dubbi e la tua storia familiare di diabete.
3.  **Non auto-diagnosticarti:** Questi sintomi possono essere causati da diverse condizioni. Solo un medico può confermare una diagnosi tramite esami del sangue specifici (come l'emoglobina glicata o il test della glicemia a digiuno).
4.  **Conosci i fattori di rischio:** Oltre ai sintomi, considera fattori come l'età, il peso e la familiarità.

**Cosa aspettarsi dal medico:**
Il medico potrebbe prescrivere:
*   **Test della glicemia a digiuno:** Misura lo zucchero nel sangue dopo una notte di digiuno.
*   **Test A1C:** Misura la media dei livelli di zucchero nel sangue negli ultimi 2-3 mesi.

In [1]:
!pip install datasets

In [2]:
from datasets import load_dataset

dataset = load_dataset("OpenMed/Medical-Reasoning-SFT-Trinity-Mini")

# Display the dataset structure
print(dataset)
# Display the first example from the training set if available
try:
    display(dataset['train'][0])
except:
    pass

README.md: 0.00B [00:00, ?B/s]

data/train-00000-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00001-of-00013.parquet:   0%|          | 0.00/226M [00:00<?, ?B/s]

data/train-00002-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00003-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00004-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00005-of-00013.parquet:   0%|          | 0.00/226M [00:00<?, ?B/s]

data/train-00006-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00007-of-00013.parquet:   0%|          | 0.00/228M [00:00<?, ?B/s]

data/train-00008-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00009-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00010-of-00013.parquet:   0%|          | 0.00/227M [00:00<?, ?B/s]

data/train-00011-of-00013.parquet:   0%|          | 0.00/226M [00:00<?, ?B/s]

data/train-00012-of-00013.parquet:   0%|          | 0.00/226M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/810284 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['messages'],
        num_rows: 810284
    })
})


{'messages': [{'content': 'Length of external auditory canal is:\n\nA.: 12 mm\nB.: 24 mm\nC.: 36 mm\nD.: 48 mm\n\nPlease reason step by step, and put your final answer within \\boxed{}.',
   'reasoning_content': None,
   'role': 'user'},
  {'content': 'The external auditory canal, also known as the external auditory meatus, is the tube that extends from the outer ear to the eardrum. In adults, the average length of this canal is approximately 2.5 cm, which is equivalent to 25 mm. Among the given options:\n\n- A: 12 mm is too short (approximately half the average length).\n- B: 24 mm is the closest to the standard adult length of 25 mm.\n- C: 36 mm is longer than the typical adult length and may correspond to other anatomical measurements, such as in infants or specific contexts, but not the standard adult canal.\n- D: 48 mm is excessively long and not representative of the external auditory canal.\n\nThus, the correct option is B.\n\n\\boxed{B}',
   'reasoning_content': 'First, the que

### Common Symptoms of Diabetes

Diabetes symptoms vary depending on how much your blood sugar is elevated. Some people, especially those with prediabetes or type 2 diabetes, may not experience symptoms initially. In type 1 diabetes, symptoms tend to come on quickly and be more severe.

**Common symptoms include:**

*   **Increased thirst (polydipsia):** Feeling thirsty more often than usual.
*   **Frequent urination (polyuria):** Needing to urinate often, particularly at night.
*   **Extreme hunger (polyphagia):** Feeling very hungry even after eating.
*   **Unexplained weight loss:** Losing weight without trying.
*   **Fatigue:** Feeling very tired and weak.
*   **Blurred vision:** Vision that isn't sharp or clear.
*   **Slow-healing sores:** Cuts or bruises that take a long time to heal.
*   **Frequent infections:** Such as gum, skin, or vaginal infections.
*   **Irritability:** Notable changes in mood.

***Important Note:** If you notice any of these symptoms, it is important to consult a healthcare professional for proper diagnosis and testing.*

# Task
Summarize the results of filtering the "OpenMed/Medical-Reasoning-SFT-Trinity-Mini" dataset for diabetes-related entries, noting that over 36,000 examples were found, and describe the specific link between Advanced Glycation End products (AGEs) and diabetes mentioned in the results. Analyze the reasoning patterns within the "messages" of this subset to identify logical steps in medical problem-solving. Finally, categorize the diabetes-related entries into sub-topics like symptoms, complications, and treatments through further filtering or clustering.

## Summarize Filtering Results

### Subtask:
Provide a summary of the diabetes-related data and describe the link between AGEs and diabetes found in the dataset.


**Reasoning**:
I will search the `diabetes_subset` for mentions of 'AGEs' or 'Advanced Glycation End products' to identify relevant examples that explain the link between hyperglycemia and tissue damage.



In [4]:
def contains_ages(example):
    # Check for 'AGEs' or 'Advanced Glycation End products' in the conversation
    for msg in example['messages']:
        text = msg['content'].lower()
        if 'ages' in text or 'advanced glycation end products' in text or 'prodotti di glicazione avanzata' in text:
            return True
    return False

# Filter the diabetes subset for AGEs
ages_subset = diabetes_subset.filter(contains_ages)

print(f'Found {len(ages_subset)} examples mentioning AGEs within the diabetes subset.')

# Display the first example to extract the explanation
if len(ages_subset) > 0:
    print('\n--- Example content regarding AGEs ---')
    print(ages_subset[0]['messages'][-1]['content'][:1500] + '...') # Displaying the assistant\'s response or relevant part
else:
    print('No specific mention of AGEs found in the current subset. Searching the broader dataset for context.')

Filter:   0%|          | 0/36042 [00:00<?, ? examples/s]

Found 6742 examples mentioning AGEs within the diabetes subset.

--- Example content regarding AGEs ---
The question asks whether advanced glycation end products (AGEs) inhibit the expression of collagen type I and III by human gingival fibroblasts. 

AGEs are known to accumulate in tissues under conditions such as diabetes and aging, and they can bind to receptors like RAGE (Receptor for AGEs) on cells, including fibroblasts. This binding activates signaling pathways that often lead to reduced collagen synthesis. Specifically, studies on human gingival fibroblasts have demonstrated that AGEs downregulate the expression of collagen genes (e.g., COL1A1 and COL3A1), resulting in decreased production of collagen type I and III. For instance, research shows that AGEs inhibit collagen synthesis through mechanisms involving oxidative stress, inflammation, and altered gene expression, which compromises tissue repair and integrity in the gingiva.

Given this evidence, the inhibition is well-su

### Summary of Diabetes-Related Data and the Role of AGEs

**1. Dataset Filtering Results:**
The initial filtering of the `OpenMed/Medical-Reasoning-SFT-Trinity-Mini` dataset identified **36,042** examples related to diabetes (using keywords 'diabetes' and 'diabete'). Within this subset, **6,742** entries specifically mention **Advanced Glycation End products (AGEs)**.

**2. The Link Between AGEs and Diabetes:**
Based on the data retrieved, the relationship between diabetes and AGEs is characterized by the following mechanism:
*   **Hyperglycemia and Formation:** Persistent high blood sugar (hyperglycemia) leads to the non-enzymatic glycation of proteins, lipids, and nucleic acids. This process results in the formation and accumulation of Advanced Glycation End products (AGEs) in various tissues.
*   **RAGE Activation:** These AGEs bind to specific receptors, such as **RAGE** (Receptor for AGEs), found on various cell types including fibroblasts.
*   **Tissue Damage:** The binding of AGEs to RAGE activates signaling pathways that trigger **oxidative stress** and **inflammation**.
*   **Impact on Repair:** Specifically, the dataset highlights that AGEs can inhibit the expression of collagen genes (Type I and III) in fibroblasts, compromising tissue repair and structural integrity. This mechanism is a key factor in the long-term complications of diabetes, such as impaired wound healing and vascular damage.

### Summary of Diabetes-Related Data and the Role of AGEs

**1. Dataset Filtering Results:**
The initial filtering of the `OpenMed/Medical-Reasoning-SFT-Trinity-Mini` dataset identified **36,042** examples related to diabetes (using keywords 'diabetes' and 'diabete'). Within this subset, **6,742** entries specifically mention **Advanced Glycation End products (AGEs)**.

**2. The Link Between AGEs and Diabetes:**
Based on the data retrieved, the relationship between diabetes and AGEs is characterized by the following mechanism:
*   **Hyperglycemia and Formation:** Persistent high blood sugar (hyperglycemia) leads to the non-enzymatic glycation of proteins, lipids, and nucleic acids. This process results in the formation and accumulation of Advanced Glycation End products (AGEs) in various tissues.
*   **RAGE Activation:** These AGEs bind to specific receptors, such as **RAGE** (Receptor for AGEs), found on various cell types including fibroblasts.
*   **Tissue Damage:** The binding of AGEs to RAGE activates signaling pathways that trigger **oxidative stress** and **inflammation**.
*   **Impact on Repair:** Specifically, the dataset highlights that AGEs can inhibit the expression of collagen genes (Type I and III) in fibroblasts, compromising tissue repair and structural integrity. This mechanism is a key factor in the long-term complications of diabetes, such as impaired wound healing and vascular damage.

## Analyze Reasoning Patterns

### Subtask:
Examine the 'messages' field in the diabetes subset to identify and describe the typical logical steps used in medical problem-solving.


**Reasoning**:
I will sample 5 examples from the `diabetes_subset` to inspect the structure of the 'messages' and identify how reasoning is presented (e.g., in a specific role or within the content).



In [5]:
import random

# Sample 5 examples from the diabetes_subset
sample_indices = random.sample(range(len(diabetes_subset)), 5)
samples = [diabetes_subset[int(i)] for i in sample_indices]

for idx, example in enumerate(samples):
    print(f'--- Sample {idx + 1} ---')
    for msg in example['messages']:
        role = msg['role']
        content = msg['content']
        # Print a snippet of the content to understand the structure
        print(f'Role: {role}')
        print(f'Content Snippet: {content[:300]}...')
    print('\n')

--- Sample 1 ---
Role: user
Content Snippet: Is type 2 diabetes phenotype and progression significantly different if diagnosed before versus after 65 years of age?

A.: yes
B.: no
C.: maybe

Please reason step by step, and put your final answer within \boxed{}....
Role: assistant
Content Snippet: The question asks whether the phenotype and progression of type 2 diabetes are significantly different based on age at diagnosis, specifically comparing diagnosis before versus after 65 years of age.

- **Phenotype Differences**: Type 2 diabetes presents differently across age groups. In individuals...


--- Sample 2 ---
Role: user
Content Snippet: Please summarize the following context:

transplant waiting list. One of the patient groups highlighted that being on dialysis can feel like "sitting and waiting and feeling like everything's on hold". The patient expert explained that although people recognise that a kidney transplant is not withou...
Role: assistant
Content Snippet: <think>
Okay, 

### Analysis of Reasoning Patterns in Diabetes-Related Entries

By examining the `messages` field in the diabetes subset, several distinct patterns in medical problem-solving and reasoning structure were identified:

**1. Reasoning Structure and Placement:**
*   **Integrated Reasoning:** In many cases, the assistant provides a step-by-step breakdown directly within the `content` of the message, often following a user request to "reason step by step."
*   **Explicit Thought Blocks:** Some examples include a `<think>` tag at the beginning of the assistant's response, where the model "thinks aloud" to plan its answer or summarize complex clinical contexts.
*   **Dedicated Metadata:** Inspection of the message objects reveals that some entries contain a specific `reasoning_content` field (distinct from the final `content`), providing a raw trace of the logical path taken.

**2. Typical Logical Steps in Medical Problem-Solving:**
*   **Identification of Clinical Variables:** The model often starts by extracting key data points from the user prompt, such as glucose levels, pH, symptoms (e.g., Kussmaul respirations), and age (e.g., diagnosis before vs. after 65).
*   **Physiological and Pathophysiological Mapping:** Reasoning frequently involves referencing specific biological mechanisms, such as the deficiency of ADH in diabetes insipidus or the role of ketone bodies in metabolic acidosis.
*   **Differential Evaluation:** For multiple-choice questions, the reasoning follows a deductive process, evaluating each option against clinical knowledge to rule out incorrect answers.
*   **Synthesis and Conclusion:** The final step involves synthesizing the analyzed variables and mechanisms to provide a diagnosis, expected lab value, or clinical recommendation, often concluding with a boxed answer.

**3. Example Snippets of Reasoning:**
*   *Variable Extraction:* "The patient presents with symptoms and laboratory findings consistent with DKA... glucose level of 380 mg/dL, and a pH of 7.3."
*   *Mechanism Explanation:* "Diabetes insipidus (DI) is characterized by the inability to concentrate urine due to a deficiency in antidiuretic hormone (ADH)..."
*   *Deductive Logic:* "Evaluating the options: a. Has no proteins (characteristic), b. Has no sugar (characteristic)... Thus, the exception is option c."

## Topic Clustering

### Subtask:
Categorize the diabetes-related entries into sub-topics like symptoms, complications, and treatments using keyword-based filtering.


## Summary:

### Q&A

**How many diabetes-related entries were identified in the dataset, and how many specifically mention Advanced Glycation End products (AGEs)?**
A total of 36,042 entries were identified as diabetes-related. Within this subset, 6,742 entries specifically mention Advanced Glycation End products (AGEs).

**What is the specific biological link between AGEs and diabetes described in the data?**
Persistent hyperglycemia leads to the non-enzymatic glycation of proteins and lipids, forming AGEs. These AGEs bind to receptors (RAGE), triggering oxidative stress and inflammation. Specifically, they inhibit the expression of collagen genes (Type I and III) in fibroblasts, which impairs tissue repair and contributes to diabetic complications like vascular damage and poor wound healing.

**What logical steps does the model follow when solving medical problems related to diabetes?**
The model typically follows a four-step reasoning process:
1.  **Clinical Variable Identification:** Extracting key data points like glucose levels, pH, and specific symptoms.
2.  **Pathophysiological Mapping:** Linking symptoms to biological mechanisms (e.g., ketone body accumulation).
3.  **Differential Evaluation:** Using deductive logic to rule out incorrect options in multiple-choice scenarios.
4.  **Synthesis:** Combining data and theory to provide a final diagnosis or recommendation.

### Data Analysis Key Findings

*   **Significant Data Volume:** The dataset contains a substantial volume of specialized medical data, with 36,042 diabetes-related examples, representing a significant portion of the "OpenMed/Medical-Reasoning-SFT-Trinity-Mini" collection.
*   **Detailed Pathological Insights:** The subset provides granular detail on diabetic complications, specifically identifying how AGEs inhibit Type I and III collagen genes, directly linking molecular biology to clinical outcomes like impaired healing.
*   **Structured Reasoning Metadata:** The analysis revealed that the dataset is highly structured for training reasoning models, utilizing integrated step-by-step content, explicit `<think>` tags, and dedicated `reasoning_content` fields to capture the logical path from symptoms to diagnosis.
*   **Deductive Methodology:** Reasoning patterns frequently exhibit a "process of elimination" strategy, where the model evaluates clinical variables against established medical knowledge to discard unlikely diagnoses.

### Insights or Next Steps

*   **Topic Clustering:** The next logical step is to perform keyword-based filtering or unsupervised clustering to further categorize the 36,042 entries into specific sub-topics such as symptoms, treatments (e.g., insulin protocols), and chronic complications.
*   **Evaluation of Reasoning Quality:** Given the presence of explicit `<think>` tags, a valuable next step would be to assess the correlation between the depth of the "thought blocks" and the accuracy of the final medical recommendations provided by the model.
